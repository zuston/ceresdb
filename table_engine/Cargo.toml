# Copyright 2023 The CeresDB Authors
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

[package]
name = "table_engine"

[package.version]
workspace = true

[package.authors]
workspace = true

[package.edition]
workspace = true

[dependencies]
# In alphabetical order
arrow = { workspace = true }
arrow_ext = { workspace = true }
async-trait = { workspace = true }
bytes_ext = { workspace = true }
ceresdbproto = { workspace = true }
common_types = { workspace = true }
datafusion = { workspace = true }
datafusion-proto = { workspace = true }
df_operator = { workspace = true }
futures = { workspace = true }
generic_error = { workspace = true }
hash_ext = { workspace = true }
itertools = { workspace = true }
lazy_static = { workspace = true }
log = { workspace = true }
macros = { workspace = true }
prometheus = { workspace = true }
prost = { workspace = true }
rand = { workspace = true }
regex = { workspace = true }
runtime = { workspace = true }
serde = { workspace = true }
smallvec = { workspace = true }
snafu = { workspace = true }
time_ext = { workspace = true }
tokio = { workspace = true }
trace_metric = { workspace = true }

[dev-dependencies]
common_types = { workspace = true, features = ["test"] }
