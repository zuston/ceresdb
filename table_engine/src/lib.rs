// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Table engine facade, provides read/write interfaces of table

pub mod engine;
pub mod memory;
pub mod metrics;
pub mod partition;
pub mod predicate;
pub mod provider;
pub mod proxy;
pub mod remote;
pub mod stream;
pub mod table;

pub const MEMORY_ENGINE_TYPE: &str = "Memory";
pub const ANALYTIC_ENGINE_TYPE: &str = "Analytic";
pub const PARTITION_TABLE_ENGINE_TYPE: &str = "PartitionTable";
