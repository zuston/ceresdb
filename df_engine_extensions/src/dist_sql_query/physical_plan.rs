// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! dist sql query physical plans

use std::{
    any::Any,
    fmt,
    pin::Pin,
    sync::Arc,
    task::{Context, Poll},
};

use arrow::{datatypes::SchemaRef as ArrowSchemaRef, record_batch::RecordBatch};
use datafusion::{
    error::{DataFusionError, Result as DfResult},
    execution::TaskContext,
    physical_expr::PhysicalSortExpr,
    physical_plan::{
        aggregates::{AggregateExec, AggregateMode},
        coalesce_batches::CoalesceBatchesExec,
        coalesce_partitions::CoalescePartitionsExec,
        displayable,
        filter::FilterExec,
        projection::ProjectionExec,
        repartition::RepartitionExec,
        DisplayAs, DisplayFormatType, ExecutionPlan, Partitioning, RecordBatchStream,
        SendableRecordBatchStream as DfSendableRecordBatchStream, Statistics,
    },
};
use futures::{future::BoxFuture, FutureExt, Stream, StreamExt};
use table_engine::{remote::model::TableIdentifier, table::ReadRequest};

use crate::dist_sql_query::RemotePhysicalPlanExecutor;

/// Placeholder of partitioned table's scan plan
/// It is inexecutable actually and just for carrying the necessary information
/// of building remote execution plans for sub tables.
// TODO: can we skip this and generate `ResolvedPartitionedScan` directly?
#[derive(Debug)]
pub struct UnresolvedPartitionedScan {
    pub sub_tables: Vec<TableIdentifier>,
    pub read_request: ReadRequest,
}

impl ExecutionPlan for UnresolvedPartitionedScan {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn schema(&self) -> ArrowSchemaRef {
        self.read_request
            .projected_schema
            .to_projected_arrow_schema()
    }

    fn output_partitioning(&self) -> Partitioning {
        Partitioning::UnknownPartitioning(self.sub_tables.len())
    }

    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
        None
    }

    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
        vec![]
    }

    fn with_new_children(
        self: Arc<Self>,
        _children: Vec<Arc<dyn ExecutionPlan>>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        Err(DataFusionError::Internal(
            "UnresolvedPartitionedScan should not have children".to_string(),
        ))
    }

    fn execute(
        &self,
        _partition: usize,
        _context: Arc<TaskContext>,
    ) -> DfResult<DfSendableRecordBatchStream> {
        Err(DataFusionError::Internal(
            "UnresolvedPartitionedScan can not be executed".to_string(),
        ))
    }

    fn statistics(&self) -> Statistics {
        Statistics::default()
    }
}

impl DisplayAs for UnresolvedPartitionedScan {
    fn fmt_as(&self, _t: DisplayFormatType, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "UnresolvedPartitionedScan: sub_tables={:?}, read_request:{:?}, partition_count={}",
            self.sub_tables,
            self.read_request,
            self.output_partitioning().partition_count(),
        )
    }
}

/// The executable scan plan of the partitioned table
/// It includes remote execution plans of sub tables, and will send them to
/// related nodes to execute.
#[derive(Debug)]
pub struct ResolvedPartitionedScan {
    pub remote_exec_ctx: Arc<RemoteExecContext>,
    pub pushing_down: bool,
}

impl ResolvedPartitionedScan {
    pub fn try_to_push_down_more(
        &self,
        cur_node: Arc<dyn ExecutionPlan>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        // Can not push more...
        if !self.pushing_down {
            return cur_node.with_new_children(vec![self.push_down_finished()]);
        }

        // Push down more, and when occur the terminated push down able node, we need to
        // set `can_push_down_more` false.
        let push_down_able_opt = PushDownAble::try_new(cur_node.clone());
        let (node, can_push_down_more) = match push_down_able_opt {
            Some(PushDownAble::Continue(node)) => (node, true),
            Some(PushDownAble::Terminated(node)) => (node, false),
            None => {
                let partitioned_scan = self.push_down_finished();
                return cur_node.with_new_children(vec![partitioned_scan]);
            }
        };

        let new_plans = self
            .remote_exec_ctx
            .plans
            .iter()
            .map(|(table, plan)| {
                node.clone()
                    .with_new_children(vec![plan.clone()])
                    .map(|extended_plan| (table.clone(), extended_plan))
            })
            .collect::<DfResult<Vec<_>>>()?;

        let remote_exec_ctx = Arc::new(RemoteExecContext {
            executor: self.remote_exec_ctx.executor.clone(),
            plans: new_plans,
        });
        let plan = ResolvedPartitionedScan {
            remote_exec_ctx,
            pushing_down: can_push_down_more,
        };

        Ok(Arc::new(plan))
    }

    pub fn new(
        remote_executor: Arc<dyn RemotePhysicalPlanExecutor>,
        remote_exec_plans: Vec<(TableIdentifier, Arc<dyn ExecutionPlan>)>,
    ) -> Self {
        let remote_exec_ctx = Arc::new(RemoteExecContext {
            executor: remote_executor,
            plans: remote_exec_plans,
        });

        Self {
            remote_exec_ctx,
            pushing_down: true,
        }
    }

    pub fn push_down_finished(&self) -> Arc<dyn ExecutionPlan> {
        Arc::new(Self {
            remote_exec_ctx: self.remote_exec_ctx.clone(),
            pushing_down: false,
        })
    }
}

#[derive(Debug)]
pub struct RemoteExecContext {
    executor: Arc<dyn RemotePhysicalPlanExecutor>,
    plans: Vec<(TableIdentifier, Arc<dyn ExecutionPlan>)>,
}

pub enum PushDownAble {
    Continue(Arc<dyn ExecutionPlan>),
    Terminated(Arc<dyn ExecutionPlan>),
}

impl PushDownAble {
    pub fn try_new(plan: Arc<dyn ExecutionPlan>) -> Option<Self> {
        if let Some(aggr) = plan.as_any().downcast_ref::<AggregateExec>() {
            if *aggr.mode() == AggregateMode::Partial {
                Some(Self::Terminated(plan))
            } else {
                None
            }
        } else if let Some(_) = plan.as_any().downcast_ref::<FilterExec>() {
            Some(Self::Continue(plan))
        } else if let Some(_) = plan.as_any().downcast_ref::<ProjectionExec>() {
            Some(Self::Continue(plan))
        } else if let Some(_) = plan.as_any().downcast_ref::<RepartitionExec>() {
            Some(Self::Continue(plan))
        } else if let Some(_) = plan.as_any().downcast_ref::<CoalescePartitionsExec>() {
            Some(Self::Continue(plan))
        } else if let Some(_) = plan.as_any().downcast_ref::<CoalesceBatchesExec>() {
            Some(Self::Continue(plan))
        } else {
            None
        }
    }
}

impl ExecutionPlan for ResolvedPartitionedScan {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn schema(&self) -> ArrowSchemaRef {
        self.remote_exec_ctx
            .plans
            .first()
            .expect("remote_exec_plans should not be empty")
            .1
            .schema()
    }

    fn output_partitioning(&self) -> Partitioning {
        Partitioning::UnknownPartitioning(self.remote_exec_ctx.plans.len())
    }

    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
        None
    }

    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
        self.remote_exec_ctx
            .plans
            .iter()
            .map(|(_, plan)| plan.clone())
            .collect()
    }

    fn with_new_children(
        self: Arc<Self>,
        _children: Vec<Arc<dyn ExecutionPlan>>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        Err(DataFusionError::Internal(
            "UnresolvedPartitionedScan can't be built directly from new children".to_string(),
        ))
    }

    fn execute(
        &self,
        partition: usize,
        context: Arc<TaskContext>,
    ) -> DfResult<DfSendableRecordBatchStream> {
        if self.pushing_down {
            return Err(DataFusionError::Internal(format!(
                "partitioned scan can't be executed before pushing down finished, plan:{}",
                displayable(self).indent(true).to_string()
            )));
        }

        let (sub_table, plan) = &self.remote_exec_ctx.plans[partition];

        // Send plan for remote execution.
        let stream_future =
            self.remote_exec_ctx
                .executor
                .execute(sub_table.clone(), &context, plan.clone())?;
        let record_stream = PartitionedScanStream::new(stream_future, plan.schema());

        Ok(Box::pin(record_stream))
    }

    fn statistics(&self) -> Statistics {
        Statistics::default()
    }
}

/// Partitioned scan stream
pub(crate) struct PartitionedScanStream {
    /// Future to init the stream
    stream_future: BoxFuture<'static, DfResult<DfSendableRecordBatchStream>>,

    /// Stream to poll the records
    stream_state: StreamState,

    /// Record schema
    arrow_record_schema: ArrowSchemaRef,
}

impl PartitionedScanStream {
    /// Create an empty RecordBatchStream
    pub fn new(
        stream_future: BoxFuture<'static, DfResult<DfSendableRecordBatchStream>>,
        arrow_record_schema: ArrowSchemaRef,
    ) -> Self {
        Self {
            stream_future,
            stream_state: StreamState::Initializing,
            arrow_record_schema,
        }
    }
}

impl RecordBatchStream for PartitionedScanStream {
    fn schema(&self) -> ArrowSchemaRef {
        self.arrow_record_schema.clone()
    }
}

impl Stream for PartitionedScanStream {
    type Item = DfResult<RecordBatch>;

    fn poll_next(self: Pin<&mut Self>, cx: &mut Context<'_>) -> Poll<Option<Self::Item>> {
        let this = self.get_mut();

        loop {
            let stream_state = &mut this.stream_state;
            match stream_state {
                StreamState::Initializing => {
                    let poll_res = this.stream_future.poll_unpin(cx);
                    match poll_res {
                        Poll::Ready(Ok(stream)) => {
                            *stream_state = StreamState::Polling(stream);
                        }
                        Poll::Ready(Err(e)) => {
                            *stream_state = StreamState::InitializeFailed;
                            return Poll::Ready(Some(Err(e)));
                        }
                        Poll::Pending => return Poll::Pending,
                    }
                }
                StreamState::InitializeFailed => return Poll::Ready(None),
                StreamState::Polling(stream) => return stream.poll_next_unpin(cx),
            }
        }
    }
}

/// Stream state
/// Before polling record batch from it, we must initializing the record batch
/// stream first. The process of state changing is like:
///
/// ```plaintext
///     ┌────────────┐                                        
///     │Initializing│                                        
///     └──────┬─────┘                                        
///   _________▽_________     ┌──────────────────────────────┐
///  ╱                   ╲    │Polling(we just return the    │
/// ╱ Success to init the ╲___│inner stream's polling result)│
/// ╲ record batch stream ╱yes└──────────────────────────────┘
///  ╲___________________╱                                    
///           │no                                            
///  ┌────────▽───────┐                                      
///  │InitializeFailed│                                      
///  └────────────────┘                                      
/// ```
pub(crate) enum StreamState {
    Initializing,
    InitializeFailed,
    Polling(DfSendableRecordBatchStream),
}

impl DisplayAs for ResolvedPartitionedScan {
    fn fmt_as(&self, t: DisplayFormatType, f: &mut fmt::Formatter) -> fmt::Result {
        match t {
            DisplayFormatType::Default | DisplayFormatType::Verbose => {
                write!(
                    f,
                    "ResolvedPartitionedScan: pushing_down:{}, partition_count:{}",
                    self.pushing_down,
                    self.remote_exec_ctx.plans.len()
                )
            }
        }
    }
}

/// Placeholder of sub table's scan plan
/// It is inexecutable actually and just for carrying the necessary information
/// of building the executable scan plan.
#[derive(Debug, Clone)]
pub struct UnresolvedSubTableScan {
    pub table: TableIdentifier,
    pub read_request: ReadRequest,
}

impl ExecutionPlan for UnresolvedSubTableScan {
    fn as_any(&self) -> &dyn Any {
        self
    }

    fn schema(&self) -> ArrowSchemaRef {
        self.read_request
            .projected_schema
            .to_projected_arrow_schema()
    }

    fn output_partitioning(&self) -> Partitioning {
        Partitioning::UnknownPartitioning(self.read_request.opts.read_parallelism)
    }

    fn output_ordering(&self) -> Option<&[PhysicalSortExpr]> {
        None
    }

    fn children(&self) -> Vec<Arc<dyn ExecutionPlan>> {
        vec![]
    }

    fn with_new_children(
        self: Arc<Self>,
        _children: Vec<Arc<dyn ExecutionPlan>>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        Err(DataFusionError::Internal(
            "UnresolvedSubTableScan should not have children".to_string(),
        ))
    }

    fn execute(
        &self,
        _partition: usize,
        _context: Arc<TaskContext>,
    ) -> DfResult<DfSendableRecordBatchStream> {
        Err(DataFusionError::Internal(
            "UnresolvedSubTableScan can not be executed".to_string(),
        ))
    }

    fn statistics(&self) -> Statistics {
        Statistics::default()
    }
}

impl DisplayAs for UnresolvedSubTableScan {
    fn fmt_as(&self, _t: DisplayFormatType, f: &mut fmt::Formatter) -> fmt::Result {
        write!(
            f,
            "UnresolvedSubTableScan: table:{:?}, request:{:?}, partition_count:{}",
            self.table,
            self.read_request,
            self.output_partitioning().partition_count(),
        )
    }
}

impl TryFrom<ceresdbproto::remote_engine::UnresolvedSubScan> for UnresolvedSubTableScan {
    type Error = DataFusionError;

    fn try_from(
        value: ceresdbproto::remote_engine::UnresolvedSubScan,
    ) -> Result<Self, Self::Error> {
        let table_ident: TableIdentifier = value
            .table
            .ok_or(DataFusionError::Internal(
                "table ident not found".to_string(),
            ))?
            .into();
        let read_request: ReadRequest = value
            .read_request
            .ok_or(DataFusionError::Internal(
                "read request not found".to_string(),
            ))?
            .try_into()
            .map_err(|e| {
                DataFusionError::Internal(format!("failed to decode read request, err:{e}"))
            })?;

        Ok(Self {
            table: table_ident,
            read_request,
        })
    }
}

impl TryFrom<UnresolvedSubTableScan> for ceresdbproto::remote_engine::UnresolvedSubScan {
    type Error = DataFusionError;

    fn try_from(value: UnresolvedSubTableScan) -> Result<Self, Self::Error> {
        let table_ident: ceresdbproto::remote_engine::TableIdentifier = value.table.into();
        let read_request: ceresdbproto::remote_engine::TableReadRequest =
            value.read_request.try_into().map_err(|e| {
                DataFusionError::Internal(format!("failed to encode read request, err:{e}"))
            })?;

        Ok(Self {
            table: Some(table_ident),
            read_request: Some(read_request),
        })
    }
}

#[cfg(test)]
mod test {
    use datafusion::error::DataFusionError;
    use futures::StreamExt;

    use crate::dist_sql_query::{
        physical_plan::PartitionedScanStream,
        test_util::{MockPartitionedScanStreamBuilder, PartitionedScanStreamCase},
    };

    #[tokio::test]
    async fn test_stream_poll_success() {
        let builder = MockPartitionedScanStreamBuilder::new(PartitionedScanStreamCase::Success);
        let mut stream = builder.build();
        let result_opt = stream.next().await;
        assert!(result_opt.is_none());
    }

    #[tokio::test]
    async fn test_stream_init_failed() {
        let builder =
            MockPartitionedScanStreamBuilder::new(PartitionedScanStreamCase::InitializeFailed);
        let stream = builder.build();
        test_stream_failed_state(stream, "failed to init").await
    }

    #[tokio::test]
    async fn test_stream_poll_failed() {
        let builder = MockPartitionedScanStreamBuilder::new(PartitionedScanStreamCase::PollFailed);
        let stream = builder.build();
        test_stream_failed_state(stream, "failed to poll").await
    }

    async fn test_stream_failed_state(mut stream: PartitionedScanStream, failed_msg: &str) {
        // Error happened, check error message.
        let result_opt = stream.next().await;
        assert!(result_opt.is_some());
        let result = result_opt.unwrap();
        assert!(result.is_err());
        let err = result.unwrap_err();
        match err {
            DataFusionError::Internal(msg) => {
                assert!(msg.contains(failed_msg))
            }
            other => panic!("unexpected error:{other}"),
        }

        // Should return `None` in next poll.
        let result_opt = stream.next().await;
        assert!(result_opt.is_none());
    }
}
