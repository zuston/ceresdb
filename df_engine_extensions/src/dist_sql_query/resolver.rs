// Copyright 2023 The CeresDB Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::sync::Arc;

use async_recursion::async_recursion;
use catalog::manager::ManagerRef as CatalogManagerRef;
use datafusion::{
    error::{DataFusionError, Result as DfResult},
    physical_plan::{
        aggregates::{AggregateExec, AggregateMode},
        ExecutionPlan,
    },
};
use table_engine::{remote::model::TableIdentifier, table::TableRef};

use crate::dist_sql_query::{
    physical_plan::{
        ResolvedPartitionedScan, SubTablePlanContext, UnresolvedPartitionedScan,
        UnresolvedSubTableScan,
    },
    ExecutableScanBuilderRef, RemotePhysicalPlanExecutorRef,
};

/// Resolver which makes datafuison dist query related plan executable.
///
/// The reason we define a `Resolver` rather than `physical optimization rule`
/// is: As I see, physical optimization rule is responsible for optimizing a bad
/// plan to good one, rather than making a inexecutable plan executable.
/// So we define `Resolver` to make it, it may be somthing similar to task
/// generator responsible for generating task for executor to run based on
/// physical plan.
pub struct Resolver {
    remote_executor: RemotePhysicalPlanExecutorRef,
    catalog_manager: CatalogManagerRef,
    scan_builder: ExecutableScanBuilderRef,
}

impl Resolver {
    pub fn new(
        remote_executor: RemotePhysicalPlanExecutorRef,
        catalog_manager: CatalogManagerRef,
        scan_builder: ExecutableScanBuilderRef,
    ) -> Self {
        Self {
            remote_executor,
            catalog_manager,
            scan_builder,
        }
    }

    /// Resolve partitioned scan
    pub fn resolve_partitioned_scan(
        &self,
        plan: Arc<dyn ExecutionPlan>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        let resolved_plan = self.resolve_partitioned_scan_internal(plan)?;

        if let Some(plan) = resolved_plan
            .as_any()
            .downcast_ref::<ResolvedPartitionedScan>()
        {
            Ok(plan.push_down_finished())
        } else {
            Ok(resolved_plan)
        }
    }

    pub fn resolve_partitioned_scan_internal(
        &self,
        plan: Arc<dyn ExecutionPlan>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        // Leave node, let's resolve it and return.
        if let Some(unresolved) = plan.as_any().downcast_ref::<UnresolvedPartitionedScan>() {
            let metrics_collector = unresolved.metrics_collector.clone();
            let sub_tables = unresolved.sub_tables.clone();
            let remote_plans = sub_tables
                .into_iter()
                .map(|table| {
                    let plan = Arc::new(UnresolvedSubTableScan {
                        table: table.clone(),
                        read_request: unresolved.read_request.clone(),
                    });
                    let sub_metrics_collect = metrics_collector.span(table.table.clone());

                    SubTablePlanContext::new(table, plan, sub_metrics_collect)
                })
                .collect::<Vec<_>>();

            return Ok(Arc::new(ResolvedPartitionedScan::new(
                self.remote_executor.clone(),
                remote_plans,
                metrics_collector,
            )));
        }

        let children = plan.children().clone();
        // Occur some node isn't table scan but without children? It should return, too.
        if children.is_empty() {
            return Ok(plan);
        }

        // Resolve children if exist.
        let mut new_children = Vec::with_capacity(children.len());
        for child in children {
            let child = self.resolve_partitioned_scan_internal(child)?;

            new_children.push(child);
        }

        Self::maybe_push_down_to_remote_plans(new_children, plan)
    }

    fn maybe_push_down_to_remote_plans(
        mut new_children: Vec<Arc<dyn ExecutionPlan>>,
        current_node: Arc<dyn ExecutionPlan>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        // No children, just return.
        if new_children.is_empty() {
            return Ok(current_node);
        }

        // This node Has multiple children, it can't be pushed down to remote.
        // But it's possible that `ResolvedPartitionedScan`s exist among its children,
        // we need to extract these children and mark them push down finished.
        if new_children.len() > 1 {
            new_children.iter_mut().for_each(|child| {
                if let Some(plan) = child.as_any().downcast_ref::<ResolvedPartitionedScan>() {
                    *child = plan.push_down_finished();
                }
            });
            return current_node.with_new_children(new_children);
        }

        // Has ensured that this node has just child and it is just
        // `ResolvedPartitionedScan`, try to push down it to remote plans in
        // `ResolvedPartitionedScan`.
        let child = new_children.first().unwrap();
        let partitioned_scan =
            if let Some(plan) = child.as_any().downcast_ref::<ResolvedPartitionedScan>() {
                plan
            } else {
                return current_node.with_new_children(new_children);
            };

        partitioned_scan.try_to_push_down_more(current_node.clone())
    }

    #[async_recursion]
    pub async fn resolve_sub_scan(
        &self,
        plan: Arc<dyn ExecutionPlan>,
    ) -> DfResult<Arc<dyn ExecutionPlan>> {
        // Leave node, let's resolve it and return.
        let build_scan_opt =
            if let Some(unresolved) = plan.as_any().downcast_ref::<UnresolvedSubTableScan>() {
                let table = self.find_table(&unresolved.table)?;
                let read_request = unresolved.read_request.clone();

                Some((table, read_request))
            } else {
                None
            };

        if let Some((table, request)) = build_scan_opt {
            return self.scan_builder.build(table, request).await;
        }

        let children = plan.children().clone();
        // Occur some node isn't table scan but without children? It should return, too.
        if children.is_empty() {
            return Ok(plan);
        }

        // Resolve children if exist.
        let mut new_children = Vec::with_capacity(children.len());
        for child in children {
            let child = self.resolve_sub_scan(child).await?;

            new_children.push(child);
        }

        plan.with_new_children(new_children)
    }

    fn find_table(&self, table_ident: &TableIdentifier) -> DfResult<TableRef> {
        let catalog = self
            .catalog_manager
            .catalog_by_name(&table_ident.catalog)
            .map_err(|e| DataFusionError::Internal(format!("failed to find catalog, err:{e}")))?
            .ok_or(DataFusionError::Internal("catalog not found".to_string()))?;

        let schema = catalog
            .schema_by_name(&table_ident.schema)
            .map_err(|e| DataFusionError::Internal(format!("failed to find schema, err:{e}")))?
            .ok_or(DataFusionError::Internal("schema not found".to_string()))?;

        schema
            .table_by_name(&table_ident.table)
            .map_err(|e| DataFusionError::Internal(format!("failed to find table, err:{e}")))?
            .ok_or(DataFusionError::Internal("table not found".to_string()))
    }
}

#[cfg(test)]
mod test {

    use datafusion::physical_plan::displayable;

    use crate::dist_sql_query::test_util::TestContext;

    #[test]
    fn test_basic_partitioned_scan() {
        let ctx = TestContext::new();
        let plan = ctx.build_basic_partitioned_table_plan();
        let resolver = ctx.resolver();
        let new_plan = displayable(resolver.resolve_partitioned_scan(plan).unwrap().as_ref())
            .indent(true)
            .to_string();
        insta::assert_snapshot!(new_plan);
    }

    #[tokio::test]
    async fn test_basic_sub_scan() {
        let ctx = TestContext::new();
        let plan = ctx.build_basic_sub_table_plan();
        let resolver = ctx.resolver();
        let new_plan = displayable(resolver.resolve_sub_scan(plan).await.unwrap().as_ref())
            .indent(true)
            .to_string();
        insta::assert_snapshot!(new_plan);
    }

    #[tokio::test]
    async fn test_unprocessed_plan() {
        let ctx = TestContext::new();
        let plan = ctx.build_unprocessed_plan();
        let resolver = ctx.resolver();

        let original_plan_display = displayable(plan.as_ref()).indent(true).to_string();

        // It should not be processed by `resolve_partitioned_scan`.
        let new_plan = resolver.resolve_partitioned_scan(plan.clone()).unwrap();

        let new_plan_display = displayable(new_plan.as_ref()).indent(true).to_string();

        assert_eq!(original_plan_display, new_plan_display);

        // It should not be processed by `resolve_sub_scan_internal`.
        let new_plan = resolver.resolve_sub_scan(plan.clone()).await.unwrap();

        let new_plan_display = displayable(new_plan.as_ref()).indent(true).to_string();

        assert_eq!(original_plan_display, new_plan_display);
    }

    #[test]
    fn test_aggr_push_down() {
        let ctx = TestContext::new();
        let plan = ctx.build_aggr_push_down_plan();
        let resolver = ctx.resolver();
        let new_plan = displayable(resolver.resolve_partitioned_scan(plan).unwrap().as_ref())
            .indent(true)
            .to_string();
        insta::assert_snapshot!(new_plan);
    }

    #[test]
    fn test_compounded_aggr_push_down() {
        let ctx = TestContext::new();
        let plan = ctx.build_compounded_aggr_push_down_plan();
        let resolver = ctx.resolver();
        let new_plan = displayable(resolver.resolve_partitioned_scan(plan).unwrap().as_ref())
            .indent(true)
            .to_string();
        insta::assert_snapshot!(new_plan);
    }

    #[test]
    fn test_node_with_multiple_partitioned_scan_children() {
        let ctx = TestContext::new();
        let plan = ctx.build_union_plan();
        let resolver = ctx.resolver();
        let new_plan = displayable(resolver.resolve_partitioned_scan(plan).unwrap().as_ref())
            .indent(true)
            .to_string();
        insta::assert_snapshot!(new_plan);
    }
}
