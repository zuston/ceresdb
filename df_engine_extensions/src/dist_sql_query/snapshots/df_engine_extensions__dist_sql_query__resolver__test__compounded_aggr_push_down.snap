---
source: df_engine_extensions/src/dist_sql_query/resolver.rs
assertion_line: 280
expression: new_plan
---
AggregateExec: mode=Final, gby=[tag1@1 as tag1, tag2@2 as tag2], aggr=[COUNT(value), COUNT(field2)]
  CoalescePartitionsExec
    ResolvedPartitionedScan: pushing_down:false, partition_count:3
      AggregateExec: mode=Partial, gby=[tag1@1 as tag1, tag2@2 as tag2], aggr=[COUNT(value), COUNT(field2)]
        ProjectionExec: expr=[time@0 as time, tag1@1 as tag1, tag2@2 as tag2, value@3 as value, field2@4 as field2]
          FilterExec: time@0 < 1691974518000 AND tag1@1 = test_tag
            UnresolvedSubTableScan: table:TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_1" }, request:ReadRequest { request_id: RequestId(42), opts: ReadOptions { batch_size: 10000, read_parallelism: 8, deadline: None }, projected: "[time,tag1,tag2,value,field2]", predicate: "[time < TimestampMillisecond(1691974518000, None) AND tag1 = Utf8(\"test_tag\")]" }, partition_count:8
      AggregateExec: mode=Partial, gby=[tag1@1 as tag1, tag2@2 as tag2], aggr=[COUNT(value), COUNT(field2)]
        ProjectionExec: expr=[time@0 as time, tag1@1 as tag1, tag2@2 as tag2, value@3 as value, field2@4 as field2]
          FilterExec: time@0 < 1691974518000 AND tag1@1 = test_tag
            UnresolvedSubTableScan: table:TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_2" }, request:ReadRequest { request_id: RequestId(42), opts: ReadOptions { batch_size: 10000, read_parallelism: 8, deadline: None }, projected: "[time,tag1,tag2,value,field2]", predicate: "[time < TimestampMillisecond(1691974518000, None) AND tag1 = Utf8(\"test_tag\")]" }, partition_count:8
      AggregateExec: mode=Partial, gby=[tag1@1 as tag1, tag2@2 as tag2], aggr=[COUNT(value), COUNT(field2)]
        ProjectionExec: expr=[time@0 as time, tag1@1 as tag1, tag2@2 as tag2, value@3 as value, field2@4 as field2]
          FilterExec: time@0 < 1691974518000 AND tag1@1 = test_tag
            UnresolvedSubTableScan: table:TableIdentifier { catalog: "test_catalog", schema: "test_schema", table: "__test_3" }, request:ReadRequest { request_id: RequestId(42), opts: ReadOptions { batch_size: 10000, read_parallelism: 8, deadline: None }, projected: "[time,tag1,tag2,value,field2]", predicate: "[time < TimestampMillisecond(1691974518000, None) AND tag1 = Utf8(\"test_tag\")]" }, partition_count:8

